demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,235 @@
+}
+
+void QueryWrapper::closeQuery() {
+    PDH_STATUS status = PdhCloseQuery(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::logic_error("The query handle to close is not valid");
+    }
+}
+
+QueryWrapper::operator PDH_HQUERY() const {
+    return query;
+}
+
+CpuMonitor::CpuMonitor() :
+    nCores{getNCores()},
+    lastEnabled{false},
+    samplesNumber{0},
+    historySize{0},
+    cpuLoadSum(nCores, 0) {}
+
+void CpuMonitor::openQuery() {
+    QueryWrapper newQueryWrapper;
+
+    PDH_STATUS status;
+    coreTimeCounters.resize(nCores);
+    for (std::size_t i = 0; i < nCores; ++i)
+    {
+        std::wstring fullCounterPath{L"\\Processor(" + std::to_wstring(i) + L")\\% Processor Time"};
+        status = PdhAddCounterW(newQueryWrapper, fullCounterPath.c_str(), 0, &coreTimeCounters[i]);
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::system_error(status, std::system_category(), "PdhAddCounter() failed");
+        }
+        status = PdhSetCounterScaleFactor(coreTimeCounters[i], -2); // scale counter to [0, 1]
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::system_error(status, std::system_category(), "PdhSetCounterScaleFactor() failed");
+        }
+    }
+    status = PdhCollectQueryData(newQueryWrapper);
+    if (ERROR_SUCCESS != status)
+    {
+        throw std::system_error(status, std::system_category(), "PdhCollectQueryData() failed");
+    }
+    QueryWrapper dummy;
+    queryWrapper = newQueryWrapper;
+    newQueryWrapper = dummy;
+}
+
+void CpuMonitor::closeQuery() {
+    queryWrapper.closeQuery();
+    coreTimeCounters.clear();
+}
+
+void CpuMonitor::setHistorySize(std::size_t size) {
+    if (0 == historySize && 0 != size) {
+        openQuery();
+    } else if (0 != historySize && 0 == size) {
+        closeQuery();
+    }
+    historySize = size;
+    std::size_t newSize = std::min(size, cpuLoadHistory.size());
+    cpuLoadHistory.erase(cpuLoadHistory.begin(), cpuLoadHistory.end() - newSize);
+}
+
+std::size_t CpuMonitor::getHistorySize() const {
+    return historySize;
+}
+
+void CpuMonitor::collectData() {
+    PDH_STATUS status;
+    status = PdhCollectQueryData(queryWrapper);
+    if (ERROR_SUCCESS != status) {
+        throw std::system_error(status, std::system_category(), "PdhCollectQueryData() failed");
+    }
+
+    PDH_FMT_COUNTERVALUE displayValue;
+    std::vector<double> cpuLoad(coreTimeCounters.size());
+    for (std::size_t i = 0; i < coreTimeCounters.size(); i++) {
+        status = PdhGetFormattedCounterValue(coreTimeCounters[i], PDH_FMT_DOUBLE, NULL, &displayValue);
+        if (ERROR_SUCCESS != status) {
+            throw std::system_error(status, std::system_category(), "PdhGetFormattedCounterValue() failed");
+        }
+        if (PDH_CSTATUS_VALID_DATA != displayValue.CStatus && PDH_CSTATUS_NEW_DATA != displayValue.CStatus) {
+            throw std::system_error(status, std::system_category(), "Error in counter data");
```

Comment:
If you're here, `status` is guaranteed to be `ERROR_SUCCESS`, so it doesn't make sense to pass it to the exception. While notionally you could use `displayValue.CStatus` instead, I don't think the CStatus values are in the same code space as Windows error codes, so it wouldn't work correctly. I think in this case you should just throw a `runtime_error`.

Reply:
updated